composer/trainer/trainer.py

```diff
@@ -403,13 +393,187 @@ def create_from_hparams(cls, hparams: TrainerHparams) -> Trainer:
 
         return trainer
 
+    @property
+    def model(self) -> BaseMosaicModel:
+        """The original model"""
+        return ddp.get_original_model(self.state.model)
+
+    @property
+    def train_dataloader(self) -> Union[DataLoader, DataloaderSpec]:
+        """The train dataloader"""
+        if self._train_split_fn is not None and self._train_device_transformation_fn is not None:
+            return DataloaderSpec(self.state.train_dataloader, self._train_device_transformation_fn,
+                                  self._train_split_fn)
+        else:
+            return self.state.train_dataloader
+
+    @train_dataloader.setter
+    def train_dataloader(self, train_dataloader: Union[DataLoader, DataloaderSpec]):
+        if isinstance(train_dataloader, DataloaderSpec):
+            self._train_device_transformation_fn = train_dataloader.device_transform_fn
+            self._train_split_fn = train_dataloader.split_fn
+            dataloader = train_dataloader.dataloader
+        else:
+            self._train_device_transformation_fn = None
+            self._train_split_fn = None
+            dataloader = train_dataloader
+        if not isinstance(dataloader, DDPDataLoader):
+            dataloader = DDPDataLoader(dataloader)
+        self.state.train_dataloader = dataloader
+
+    # TODO(anis) -- add getters/setters for evaluators
+
+    @property
+    def max_epochs(self):
+        """Maximum number of training epochs"""
+        return self.state.max_epochs
+
+    @max_epochs.setter
+    def max_epochs(self, max_epochs: int):
+        self.state.max_epochs = max_epochs
+
+    @property
+    def algorithms(self):
+        """Algorithms"""
+        return self.state.algorithms
+
+    @algorithms.setter
+    def algorithms(self, algorithms: Sequence[Algorithm]):
+        self.state.algorithms = algorithms
+
+    @property
+    def callbacks(self):
+        """Callbacks"""
+        return self.state.callbacks
+
+    @callbacks.setter
+    def callbacks(self, callbacks: Sequence[Callback]):
+        # Preserve the logger backends as callbacks
+        self.state.callbacks = [*self.logger.backends, *callbacks]
+
+    @property
+    def optimizers(self):
+        """Optimizers"""
+        return self.state.optimizers
+
+    @optimizers.setter
+    def optimizers(self, optimizers: Optimizers):
+        self.state.optimizers = optimizers
+
+    @property
+    def schedulers(self):
+        """Schedulers"""
+        return self.state.schedulers
+
+    @schedulers.setter
+    def schedulers(self, schedulers: Schedulers):
+        self.state.schedulers = schedulers
+
+    @property
+    def device(self):
+        """Device"""
+        # No setter for device since it wouldn't make sense to change it
```

Comment:
Might be worth raising an error on setting with a pointer that the user should re-initialize the trainer if they want to switch devices.